mantle/kola/harness.go

```diff
@@ -943,6 +944,24 @@ func registerExternalTest(testname, executable, dependencydir string, userdata *
 		targetMeta = &metaCopy
 	}
 
+	switch targetMeta.Isolation {
+	case "readonly":
+	case "dynamicuser":
+		// These tests cannot provide their own Ignition
+		if userdata != nil {
+			return fmt.Errorf("test %v specifies isolation=%v but includes an Ignition config", testname, targetMeta.Isolation)
+		}
+		targetMeta.Exclusive = false
```

Comment:
```suggestion
```

Still need to drop this bit as discussed in https://github.com/coreos/coreos-assembler/pull/3060/files#r960684553 and https://github.com/coreos/coreos-assembler/pull/3060/files#discussion_r960795802

Reply:
There's no use case for `exclusive: true; isolation=<something>` tests right? 

I definitely see this *replacing* most of our use of `exclusive: false` and so I hence don't see the value in requiring it to also be specified.

Reply:
I'm just stating and restating and stating again that I don't want to imply `exclusive: false`. 

If someone added a test that was likely to cause flakes because it's run with other tests we'd catch that today because we automatically re-run failed tests on their own and we'd see that the re-run passed.

I get what you were trying to do with this, but the model we developed doesn't apply `1:1` with what you wanted it to be.

If we change anything what I would prefer is for non-exclusive tests to default to `ProtectSystem=strict` and then allow adding a tag (or I guess it could be a toplevel option) to disable that (for example to those two tests that you found that don't comply). Then a user has to think about it when they choose to override that default behavior.

Reply:
I already wrote up a comment in the other thread before I saw Dusty's reply proposing essentially the same thing.  Moving it here for completeness:

I think the isolation functionality is potentially useful, but if the goal is to reduce boilerplate comments, the proposed flags are counterproductive.  Tests have to opt in to this functionality by writing something like `"isolation": "dynamicuser"`, and if someone wants to find out what that means, they can look it up in cosa docs and learn that it means `DynamicUser=yes`.  Okay, now they have to find an option in systemd docs, which isn't easy unless they know about `systemd.directives(7)`.  And why does the test use that?  What are the implications?  Why would I select `dynamicuser` rather than `readonly`, and why can't I select both?  (`dynamicuser` implies `readonly`, but that's more spelunking in systemd docs.)

So let's figure out what we actually want here, without recreating systemd security option spaghetti.  Proposal: for tests that have `exclusive: false` and do not provide an Ignition/Butane config, automatically set `DynamicUser=yes` unless a second option is specified, something like `restrict: false`.  That way we default to maximum isolation unless the test says that it wants less; we'll easily notice that case because the test will fail.  If there are known cases where we want `ProtectSystem=strict` without `DynamicUser=yes`, we can provide an option for that too, or make the first one multi-valued.  Maybe in the Ignition + `exclusive: false` case we'd want to require an explicit `restrict: false`, to make it clear that protections are being disabled.

Reply:
> I'm just stating and restating and stating again that I don't want to imply exclusive: false.

Your seems to overlap with bgilbert's, so replying below:

> and if someone wants to find out what that means, they can look it up in cosa docs and learn that it means DynamicUser=yes. Okay, now they have to find an option in systemd docs, which isn't easy unless they know about systemd.directives(7).

Right, but the developers on the team already have to read the external-tests.md documentation to understand much at all.  Needing that isn't...isolated...to this change.

> Proposal: for tests that have exclusive: false and do not provide an Ignition/Butane config, automatically set DynamicUser=yes unless a second option is specified, something like restrict: false.

Hmmm...definitely implicit-over-explicit going on here.  I think having an ignition/butane config provided turning off isolation would be too magical, and instead we should error out and require `isolation: none` or so.

Overall I'm absolutely fine with defaulting `exclusive: false` tests to being isolated...but it will require some ratcheting.

A tricky thing here is that some of our read only tests actually do need to still run as root because they traverse the filesystem and we have some protected directories. 

This is why I introduced both `isolation: readonly` (runs as root, but read only) and `isolation: dynamicuser`.

Anyways then so then we're agreeing to:

- add `isolation: none|readonly|dynamicuser`
- Do a PR to fedora-coreos-config which adds `exclusive: false; isolation: none` to the tests that need it
- Another (or same) PR to fedora-coreos-config which adds `exclusive: false; isolation: readonly` to tests that need to be root but are still readonly
- Change this PR to default `exclusive: false` to `isolation: dynamicuser`

I need to more carefully look at the case of external tests with ignition configs and see the impact.

Reply:
my proposal is that we

- leave `dynamicuser` out of this, what does it really give us?
- default to `isolation: readonly` for all `exclusive: false` tests, including ones that provide Ignition configs. Only the test is run with isolation, obviously. 
- add `isolation: none` to tests that need r/w access to the filesystem and explain the rationale for doing so in the comments in the test

The above proposal is straightforward, has clear benefits, doesn't require option changes to a bunch of our tests, and isn't confusing.

Reply:
> leave dynamicuser out of this, what does it really give us?

It's a *much* stronger level of isolation.  Today for example this will halt the machine:

```
[root@cosa-devsh ~]# systemd-run -P -p ProtectSystem=strict halt
Running as unit: run-u36.service
[DONE] QEMU instance terminated
```

It's because `ProtectSystem=strict` just affects mounts - direct filesystem access as well as some system calls.  But...the big "escape hatch" is that doing IPC to Unix domain sockets isn't restricted.  So a `ProtectSystem=strict` unit can e.g. still itself talk to systemd as root.

In contrast, a `DynamicUser=yes` process really drops all privileges and hence e.g. if it tries to reboot the system by asking systemd to do it over IPC, will be rejected because it has a nonzero uid.

Reply:
This is really relevant to us not just for systemd, but also bootupd and rpm-ostree also work exactly the same way:

```
[root@cosa-devsh ~]# systemd-run -P -p ProtectSystem=strict bootupctl update
Running as unit: run-u28.service
No update available for any component.
[root@cosa-devsh ~]# systemd-run -P -p DynamicUser=yes bootupctl update
Running as unit: run-u29.service
error: connecting to /run/bootupd.sock: EACCES: Permission denied
[root@cosa-devsh ~]# systemd-run -P -p ProtectSystem=strict rpm-ostree install -C --dry-run cowsay
Running as unit: run-u36.service
Checking out tree 90fa26d... done
error: Packages not found: cowsay
[root@cosa-devsh ~]# systemd-run -P -p DynamicUser=yes rpm-ostree install -C --dry-run cowsay
Running as unit: run-u39.service
Error creating textual authentication agent: Error opening current controlling terminal for the process (`/dev/tty'): No such device or address (polkit-error-quark, 0)
error: rpmostreed OS operation PkgChange not allowed for user
[root@cosa-devsh ~]# 

```

Reply:
OK and also docker of course:

```
[root@cosa-devsh ~]# systemd-run -P -p ProtectSystem=strict docker run busybox echo hello
Running as unit: run-u42.service
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
2c39bef88607: Pull complete 
Digest: sha256:20142e89dab967c01765b0aea3be4cec3a5957cc330f061e5503ef6168ae6613
Status: Downloaded newer image for busybox:latest
hello
[root@cosa-devsh ~]# systemd-run -P -p DynamicUser=yes docker run busybox echo hello
Running as unit: run-u50.service
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
[root@cosa-devsh ~]# 
```

podman of course is all about being "daemonless", so it will fail to do writes with `ProtectSystem=strict`.  (Also, hmm it seems to fail with `DynamicUser=yes`...that seems like a bug potentially)

Reply:
I get the point that `DynamicUser=yes` provides grater protection, but is it worth the extra complexity? After all what we're doing here is policing our **tests** on throw away systems. What is the real value?

Reply:
> and if someone wants to find out what that means, they can look it up in cosa docs and learn that it means DynamicUser=yes. Okay, now they have to find an option in systemd docs, which isn't easy unless they know about systemd.directives(7). And why does the test use that? What are the implications?

I would expect most developers working on CoreOS to know this stuff - or at least, to know where to look for further docs!  In the same way we need to know containers, clouds, etc. a working knowledge of systemd is pretty essential.  Stated another way, that doesn't mean that you know precisely what `DynamicUser=yes` does - but it *does* mean you know about the systemd docs.  And of course, you have friendly teammates who are happy to help!

Sure, this means that more people writing tests would need to understand systemd isolation, but...for example if you touch rpm-ostree today, you need to understand it for that too: https://github.com/coreos/rpm-ostree/pull/3937

The systemd isolation stuff is definitely something we should be making more use of in the services we write and maintain.

Reply:
new proposal:

- add `isolation:` toplevel config option
    - accepts `none` or `strict`, which means `DynamicUser=yes` and `ProtectSystem=strict`
- default to `isolation: strict` for all `exclusive: false` tests, including ones that provide Ignition configs. Only the test itself is run with isolation, obviously. 
- add `isolation: none` to `exclusive: false` tests that need r/w access to the filesystem and explain the rationale for doing so in the comments in the test

Reply:
> After all what we're doing here is policing our tests on throw away systems. What is the real value?

It makes sure that anyone writing an `exclusive: false` test isn't e.g. running a privileged docker container which could affect the system, installing packages, mutating the bootloader state...so that avoids any flakes inherent.

I think about 75% of the `exclusive: false` tests can run as `DynamicUser=yes`.

And more generally, we should avoid running code as root *globally*!

Do you really think having a tristate `isolation: none|readonly|dynamicuser` is too complicated?   It doesn't seem so to me.  But I'm also OK simplifying it to `isolation: none|user` as a baseline.

We could just expose the ability to inject systemd unit directives directly:

```
  kola:
    unit: |
      [Service]
      ProtectSystem=strict
      PrivateTmp=yes
```

for those fewer tests that need root-but-readonly.

Reply:
> Do you really think having a tristate `isolation: none|readonly|dynamicuser` is too complicated?

I'm not opposed to that, but what I am opposed to is implying `exclusive: false` in any way. So i've tried to propose options that give you what you want (stricter permissions) with what we already have today.

I honestly think this is a made up problem especially since we are the ones writing our own tests and we code review them and if there were flakes due to these issues we'd see them immediately.

Reply:
While I agree that this is a useful hardening step to avoid a class of bugs in non-exclusive tests, it's not solving an urgent problem and shouldn't be a high priority.  I'm concerned about the amount of developer energy this review has already consumed.  Colin, you have two colleagues expressing concerns about this PR.  Please take a step back and reconsider your approach.

It's important to note that the people writing (or reading!) tests are often _not_ experienced developers.  They're relatively new team members who are already overwhelmed by the amount of complexity they've been handed, or they're external contributors trying to do a good deed.  The "people should know this stuff" argument simply does not work here.  That's why I think it's important to have reasonable defaults and to define the options in a way that makes it clear what's going on.  It's not at all obvious that `isolation: readonly` means "this test needs root access", and the result will be that Dusty will correctly insist that every use of the flag gets one of those comments you're opposed to.

Reply:
I definitely agree this is not urgent.  I probably should have *started* with https://github.com/coreos/coreos-assembler/pull/3064/commits/242e7149b8d57b5837238366bfcd70bd99b455a5

But I also think this is a micro-example of where we own the code and process and we can work smarter, not harder.

> It's important to note that the people writing (or reading!) tests are often not experienced developers. They're relatively new team members who are already overwhelmed by the amount of complexity they've been handed, or they're external contributors trying to do a good deed.

I understand what you're saying, but I am **extremely doubtful** that one more semantic kola tag is going to be the "straw that breaks the camel's back".  To even get to the point of running kola tests one needs to understand a whole world of stuff that's part of (Fedora) CoreOS and coreos-assembler.

> It's not at all obvious that isolation: readonly means "this test needs root access", and the result will be that Dusty will correctly insist that every use of the flag gets one of those comments you're opposed to.

I'm not opposed to comments - I'm opposed to mindlessly copy/pasting comments and requiring those to merge a PR adding a test.  There's actually quite few tests that would use `isolation: readonly` so sure, by all means they could use a comment!

Reply:
I agree that we own the code and the process.  In particular, we have complete freedom to define the new options in a way that isn't confusing, so please fix the problems with the PR rather than continuing to try to justify them.  We should default to the behavior (hardening) that we want to be the default, keyed off the existing `exclusive: false` flag, and we should define the override flags so that they clearly specify what the flags do (loosen particular restrictions) rather than what they don't do (leave certain restrictions unchanged).  I'll take another look at this PR once that is done.

Reply:
>  In particular, we have complete freedom to define the new options in a way that isn't confusing, so please fix the problems with the PR rather than continuing to try to justify them.

Hmm; sorry, I still don't think it's confusing.  

>  We should default to the behavior (hardening) that we want to be the default

Now as stated I do agree with that *but* it will require some ratcheting.  Did you see my comment here https://github.com/coreos/coreos-assembler/pull/3060#discussion_r961607099 ?  Do you agree with that?

Reply:
Proposal:

- add `isolation: none|readonly|dynamicuser` to coreos-assembler but *don't* change its relationship with `exclusive` (a small tweak to this PR)
- Do a PR to fedora-coreos-config which adds `exclusive: false; isolation: none` to the tests that need it
- Another (or same) PR to fedora-coreos-config which adds `exclusive: false; isolation: readonly` to tests that need to be root but are still readonly
- Change this PR to default `exclusive: false` to `isolation: dynamicuser`

I need to more carefully look at the case of external tests with ignition configs and see the impact.

Reply:
I did see it.  It does not address my concerns from the comment https://github.com/coreos/coreos-assembler/pull/3060#discussion_r961515491 it was replying to, nor my restatement just now in https://github.com/coreos/coreos-assembler/pull/3060#discussion_r963933261.  My -1 will remain in place until those concerns are addressed.

Reply:
Ah...you're proposing `s/isolation: none/restrict: false/` basically, right?

Reply:
I think `restrict` is slightly better than `isolation`, but the main point is avoiding `dynamicuser` and `readonly`.  Instead of `dynamicuser`, we can have something like `restrict: true` (implicit because it's the default), and instead of `readonly`, something like an implicit `restrict: true` with an explicit `run-as-root: true`.  (I'm open to better naming ideas also.)

Reply:
Err...I was with you on the first half of the sentence - sure, we can have a boolean instead a tristate (I still don't think a tristate is hard or complicated but whatever, I was mainly on board with just having a boolean).  That seems to be what Dusty was arguing for too.

Then in the second half you made up another boolean - and now we have two booleans which is I hope you'd agree if you step back for a moment, not actually simpler. 

For one thing, two booleans is *four* potential states.  The new one of which (restrict: false, run-as-root: true) is redundant.  And any future expansion via additional booleans becomes quickly grows the state space.

I think I'd much prefer just a boolean (if that's what we're going down to) alongside the ability for tests to inject arbitrary systemd unit directives, which would likely help solve other things too (e.g. ordering against other units).

Reply:
The extra boolean isn't simpler for the implementation, but we should prioritize building intuitive directives for the test author.  I think distinct flags help with that, but I'm okay with a three-valued enum if we can find sensible names for the values.  I see how the arbitrary directive injection could be useful, but using it to override `DynamicUser` would be pretty awkward, and I don't think encouraging that helps with the goal of simplifying tests.

For completeness, I'll note a couple things.  One is that an enum implies that there's only one dimension of hardening, and down the road we might find that we want to compose multiple hardening flags anyway.  The other is that systemd has the sort of interlocking options you're arguing against.  `DynamicUser` is a separate option from `ProtectSystem`, but implies it.  (systemd has a gazillion options though, which I agree can be hard to work with.)

I have deadlines I need to focus on, so I'm muting this thread for now.  I've tried to fully capture my constraints in https://github.com/coreos/coreos-assembler/pull/3060#discussion_r963933261 (but `s/flags/options/` if you want to pursue the enum).  I'll check back later, which might be after OCP feature freeze, and my -1 stands until then.